### PR TITLE
PayPal reconciliation CRON job

### PR DESCRIPTION
**Pull request by Betree**
Resolve https://github.com/opencollective/opencollective/issues/6342

To do as follow-up:
- Record refunded transactions
- Reconcile PayPal checkout orders
- Reconcile "Pass payment" (payouts)

### CODE REVIEW

File: cron/daily/51-synchronize-paypal-ledger.ts [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+  'T0002',
+  // PayPal checkout. These ones are now necessarily recorded in the database, but before we moved to "AUTHORIZE" intent
+  // instead of "CAPTURE" (in https://github.com/opencollective/opencollective-frontend/pull/8601) we got some cases where
+  // the transactions were never recorded. We'll iterate on these in a follow-up PR.
+  // 'T0006',
+];
+
+// Ignore some hosts, usually because they haven't enabled transactions search API yet
+const IGNORED_HOSTS = [
+  // Transactions search not enabled
```
Comment (Betree) [inline cron/daily/51-synchronize-paypal-ledger.ts:39]: Working on enabling them in https://github.com/opencollective/opencollective/issues/6392

File: scripts/paypal/info.ts [hunk 2]
```diff
@@ -73,6 +74,20 @@ const printAllHostsWithPaypalAccounts = async () => {
...
+const showCaptureInfo = async (hostSlug, paypalTransactionId) => {
+  const host = await models.Collective.findBySlug(hostSlug);
+  const captureUrl = `payments/captures/${paypalTransactionId}`;
+  const captureDetails = (await paypalRequestV2(captureUrl, host, 'GET')) as PaypalCapture;
+  const dbTransactions = await models.Transaction.findAll({
+    where: { data: { paypalCaptureId: paypalTransactionId } },
+  });
+
+  console.log('==== Capture details ====');
+  console.log(captureDetails);
```
Comment (kewitz) [inline scripts/paypal/info.ts:86]:
```suggestion
  console.dir(captureDetails, { depth: 10 });
```

File: scripts/paypal/payment-reconciliator.ts [hunk 3]
```diff
@@ -242,60 +232,7 @@ const reconcileSubscription = async (paypalSubscriptionId: string, _, commander)
...
-    !requiredAssociations.every(association => !order[association].deletedAt)
-  ) {
-    logger.error(
-      `Subscription ${subscription.id} has deleted entities, please restore them first: ${requiredAssociations.map(
-        association => `${association}: ${Boolean(order[association])}`,
-      )}`,
-    );
-    return;
-  }
+  const { subscription, order, host } = await loadDataForSubscription(paypalSubscriptionId);
```
Comment (kewitz) [inline scripts/paypal/payment-reconciliator.ts:235]: 🍨 Calling before definition, I would keep helper functions above.

File: scripts/paypal/payment-reconciliator.ts [hunk 4]
```diff
@@ -530,43 +467,108 @@ const findMissingPaypalTransactions = async (_, commander) => {
       totalPages = <number>response['totalPages'];
```
Comment (kewitz) [inline scripts/paypal/payment-reconciliator.ts:467]: Should we use the newly implemented `listPayPalTransactions`?

File: scripts/paypal/payment-reconciliator.ts [hunk 5]
```diff
@@ -530,43 +467,108 @@ const findMissingPaypalTransactions = async (_, commander) => {
...
-              continue; // Make sure the capture is not pending
-            }
+          // Make sure it's happening on a subscription
+          console.warn(`Missing PayPal transaction ${paypalTransactionId} in ledger`);
+          if (transactionInfo['paypal_reference_id_type'] !== 'SUB') {
+            console.error(
+              `Found a missing transaction that is not a subscription: ${paypalTransactionId}`,
+              paypalTransaction,
+            );
+            throw new Error('This case is not supported'); // Shouldn't happen, just in case
```
Comment (kewitz) [inline scripts/paypal/payment-reconciliator.ts:494]: Shouldn't we `continue` here?

File: cron/daily/51-synchronize-paypal-ledger.ts [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+import { reportErrorToSentry } from '../../server/lib/sentry';
+import { parseToBoolean } from '../../server/lib/utils';
+import models, { sequelize } from '../../server/models';
+import { paypalRequestV2 } from '../../server/paymentProviders/paypal/api';
+import { recordPaypalCapture } from '../../server/paymentProviders/paypal/payment';
+import { PaypalCapture, PaypalTransactionSearchResult } from '../../server/types/paypal';
+
+const LIMITED_TO_HOST_SLUGS = process.env.HOST ? process.env.HOST.split(',') : null;
+const START_DATE = process.env.START_DATE ? moment(process.env.START_DATE) : moment().subtract(2, 'day');
+const END_DATE = process.env.END_DATE ? moment(process.env.END_DATE) : moment(START_DATE).add(1, 'day');
```
Comment (kewitz) [inline cron/daily/51-synchronize-paypal-ledger.ts:23]: Should we use UTC dates by default?

**Review by kewitz — APPROVED** [review]
Looks good. Hopefully, this won't find many unsynced transactions since the last refactor.
I added a few notes and suggestions, but I don't think they're very relevant when considering the complexity of this PR.

File: scripts/paypal/payment-reconciliator.ts (hunk 4 above)
Comment (Betree) [reply scripts/paypal/payment-reconciliator.ts:467]: Ideally, yes. In practice, I expect this script to become mostly useless with the new CRON job, which is why I didn't invest more time updating it.

File: scripts/paypal/payment-reconciliator.ts (hunk 5 above)
Comment (Betree) [reply scripts/paypal/payment-reconciliator.ts:494]: I want to make sure we spot the error right away if this ever happens. And as mentioned, this script is probably going away (at least partially) in the near future

File: cron/daily/51-synchronize-paypal-ledger.ts (hunk 6 above)
Comment (Betree) [reply cron/daily/51-synchronize-paypal-ledger.ts:23]: Good call

Comment (Betree) [reply cron/daily/51-synchronize-paypal-ledger.ts:23]: Good call

**Pull request merged**